lib/segment/src/vector_storage/query/reco_query.rs

```diff
@@ -37,42 +37,89 @@ impl<T, U> TransformInto<RecoQuery<U>, T, U> for RecoQuery<T> {
     }
 }
 
-impl<T> Query<T> for RecoQuery<T> {
+#[derive(Debug, Clone, PartialEq)]
+pub struct RecoBestScoreQuery<T>(RecoQuery<T>);
+
+impl<T> From<RecoQuery<T>> for RecoBestScoreQuery<T> {
+    fn from(query: RecoQuery<T>) -> Self {
+        Self(query)
+    }
+}
+
+impl<T, U> TransformInto<RecoBestScoreQuery<U>, T, U> for RecoBestScoreQuery<T> {
+    fn transform<F>(self, f: F) -> OperationResult<RecoBestScoreQuery<U>>
+    where
+        F: FnMut(T) -> OperationResult<U>,
+    {
+        Ok(RecoBestScoreQuery(self.0.transform(f)?))
+    }
+}
+
+impl From<RecoBestScoreQuery<VectorInternal>> for QueryVector {
+    fn from(query: RecoBestScoreQuery<VectorInternal>) -> Self {
+        QueryVector::RecommendBestScore(query.0)
+    }
+}
+
+impl<T> Query<T> for RecoBestScoreQuery<T> {
     fn score_by(&self, similarity: impl Fn(&T) -> ScoreType) -> ScoreType {
         // get similarities to all positives
-        let positive_similarities = self.positives.iter().map(&similarity);
+        let positive_similarities = self.0.positives.iter().map(&similarity);
 
         // and all negatives
-        let negative_similarities = self.negatives.iter().map(&similarity);
+        let negative_similarities = self.0.negatives.iter().map(&similarity);
+
+        // get max similarity to positives and max to negatives
+        let max_positive = positive_similarities
+            .max_by(|a, b| a.total_cmp(b))
+            .unwrap_or(ScoreType::NEG_INFINITY);
 
-        merge_similarities(positive_similarities, negative_similarities)
+        let max_negative = negative_similarities
+            .max_by(|a, b| a.total_cmp(b))
+            .unwrap_or(ScoreType::NEG_INFINITY);
```

Comment:
Is this indeed correct or should it be the inverse like this?

```suggestion
        let max_negative = negative_similarities
            .min_by(|a, b| a.total_cmp(b))
            .unwrap_or(ScoreType::INFINITY);
```

Reply:
It is not min_by, because we are calculating whether the closest example is a negative or positive example. This means we want to know the highest similarity in both cases